dataprocspawner/spawner.py

```diff
@@ -286,20 +290,127 @@ async def stop(self):
 
   async def poll(self):
     status = await self.get_cluster_status(self.clustername())
-    if status is None or status in (ClusterStatus.State.ERROR,
-                                    ClusterStatus.State.DELETING,
+    if status is None or status in (ClusterStatus.State.DELETING,
                                     ClusterStatus.State.UNKNOWN):
       return 1
+    elif status == ClusterStatus.State.ERROR:
+      return 1
     elif status == ClusterStatus.State.CREATING:
       self.log.info(f'{self.clustername()} is creating')
       return None
     elif status in (ClusterStatus.State.RUNNING, ClusterStatus.State.UPDATING):
       self.log.info(f'{self.clustername()} is up and running')
       return None
 
-################################################################################
-# User form functions
-################################################################################
+  @async_generator
+  async def progress(self):
+    """ Loads bars progressively and displays cluster logs.
+
+    Uses Cloud Logging method names to define a list of logs to display on the
+    progress page. Those methods frames how many steps the progress bar takes
+    before the cluster creation operation is done.
+
+    If the operation is succesful, the page redirects to the notebook as usual
+    with JupyterHub.
+
+    If the operation failes, the progress bar is red and the user remains on the
+    page. Error logs are displayed on the page.
+    """
+    # Displays operation detail
+    if not self.operation:
+      msg_existing = (
+        'Trying to load progress but no cluster being created. One reason '
+         f'might be that a cluster named {self.clustername()} already exists '
+         'and it was not spawned from this Dataproc Hub instance.')
+
+      await yield_({'progress': 100, 'failed': True, 'message': msg_existing})
+      raise RuntimeError(msg_existing)
+
+    operation_id = self.operation.operation.name.split('/')[-1]
+    cluster_uuid = self.operation.metadata.cluster_uuid
+    progress = 5
+    html_message = (
+      f'Operation {operation_id} is creating cluster with uuid {cluster_uuid}')
+    await yield_({'progress': progress, 'html_message': html_message})
+
+    # Displays progress and logs using `methods` in Cloud Logging
+    logging_client = logging_v2.LoggingServiceV2Client()
+    log_methods = {
+      'doStart',
+      'instantiateMe',
+      'getOrCreateAgent',
+      'run',
+      'runBuiltinInitializationActions',
+      'awaitNameNodeSafeModeExit',
+      'runCustomInitializationActions'
+    }
+    log_size = len(log_methods)
+    log_added = []
```

Comment:
Since we always convert this into a set before using it, can we just instantiate this as a set to begin with?

Reply:
Replaced in commit fecf03cb0d40cced758907bd1d401f7e6b3bcb3d